Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,817 @@
+Alex will read a `.x` file created by us specifying how to match lexemes and it will in turn create a `.hs` file which is the generated lexer.
+
+In the second part of this tutorial, we will use Happy to generate a parser that will consume the token stream.
+Happy will try to match tokens according to specific rules that we will describe and produce a tree structure representing a valid MiniML program.
+
+Note that using Alex is not a requirement. You could also use, for example, parser combinators such as Megaparsec for lexing and Happy for parsing if you wanted to.
+
+Making a lexer using parser combinators is pretty doable and manageable.
+Meanwhile, using Alex is a bit more involved, but it has the advantage that performance will be more predictable; besides that, it can be easily integrated with Happy to output one token at a time, which may avoid creating a massive list of tokens in the memory in the first place.
+
+## Creating the project
+
+We will use Stack to build the project.
+Alex and Happy use files with extensions `.x` and `.y,` respectively, and Stack can automatically detect these file formats and build them.
+
+n.b.: Using Stack is not strictly necessary. You may also use Cabal or plain Alex and Happy files, and it will work as well.
+
+Begin by creating an empty Stack project:
+
+```bash
+$ stack new mini-ml simple
+$ cd mini-ml
+```
+
+You will also need to have Alex and Happy installed on your machine:
+
+```bash
+$ stack install alex
+$ stack install happy
+```
+
+We will create a `package.yaml` file for our project configuration.
+Use your favorite text editor to initialize `package.yaml` with the following:
+
+```yaml
+name: mini-ml
+version: 0.1.0.0
+
+extra-source-files:
+- README.md
+- LICENSE
+
+default-extensions:
+- OverloadedStrings
+
+library:
+  source-dirs: src
+  dependencies:
+  - array
+  - base
+  - bytestring
+  build-tools:
+  - alex
+  - happy
+```
+
+n.b.: If you chose to use Stack, you might need to use `hpack --force` to override the current Cabal file.
+
+You may want to remove `src/Main.hs`, as we won't need it.
+
+## Alex
+
+<!--
+* [X] link to its user guide
+-->
+
+Alex is a Haskell tool to generate lexers.
+
+This section aims at giving an introduction to Alex and how you can use it to do useful things in conjunction with Happy (explained later).
+Although Alex and Happy are frequently used together, they are independent tools. They may be combined with other technologies, so you may, for example, use Alex and Megaparsec together if you prefer.
+
+### How it works
+
+<!--
+* [X] make a nifty flowchart
+-->
+
+It will be essential to differentiate between the terms **lexeme** and **token** during this section.
+A lexeme is a valid atom of the grammar, such as a keyword (`in`, `if`, etc.), an operator (`+`, `/`, etc.), an integer literal, a string literal, a left or right parenthesis, an identifier, etc.
```

Comment:
A lexeme is something you find in the un-parsed code, while a token is something you find in parsed code -- you parse lexemes into tokens. 

Would this distinction be true and meaningful? I had to struggle a bit to grasp the difference from this paragraph.

Reply:
> A lexeme is something you find in the un-lexed code, while a token is something you find in lexed code -- you "lex" lexemes into tokens.

Eh, the last part ("lex" lexemes into tokens) is _kind of_ correct, hence lexer is also called "tokenizer", but this grinds my gears a little.

Lexeme is a fancy word for "word", i.e. a lexical unit, an element of the [lexis](https://en.wikipedia.org/wiki/Lexis_(linguistics)) (i.e. vocabulary), something you can meaningfully categorize within your grammar; natural language examples are: "cat" is a lexeme of class "noun", "an" is a lexeme of class "indefinite article", etc. In computer science we use it kind of vaguely since we often include punctuation into the lexis, which linguists don't usually do.

A token is a representation of a lexeme used internally by lexer and parser, generally agreed to consist of a _token name_, a tag identifying the corresponding lexeme's category, and _token value_, which is arbitrary data (implementation defined, but often it's just a string that contains the lexeme itself, or some other representation thereof).

These two exist in somewhat parallel realities, and certainly lexemes don't "disappear" after lexing, we just represent them as tokens in the code.